File: src/main/java/phaserush/Mar2022/Q1.java
package phaserush.Mar2022;

import java.io.File;
import java.io.FileNotFoundException;
import java.net.URISyntaxException;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Scanner;
import java.util.stream.Collectors;

public class Q1 {
    private static final Map<String, Integer> dict = new LinkedHashMap<>();

    static {
        dict.put("A", 0);
        dict.put("B", 1);
        dict.put("C", 2);
        dict.put("AB", 3);
        dict.put("BA", 4);
    }

    public static void main(String[] args) throws URISyntaxException, FileNotFoundException {
        var sc = new Scanner(new File(Q1.class.getResource("/Mar2022/q1.txt").toURI()));
        var input = sc.nextLine();
        System.out.println("Entropy:\t" + calcEntropy(input));
        var output = compress(input);
        System.out.println("Compressed:\t" + output);
        System.out.println("Ratio: \t" + ((double)input.length()/output.length()));
        System.out.println("Dictionary:");
        dict.forEach((k, v) -> System.out.println(v + "\t" + k));
    }

    private static double calcEntropy(final String input) {
        var freq = input.chars()
                .mapToObj(c -> (char) c)
                .collect(Collectors.groupingBy(i -> i, Collectors.counting()));
        return freq.keySet().stream()
                .mapToDouble(ch -> ((double) freq.get(ch)) / input.length())
                .reduce(0d, (acc, curr) -> acc - curr * Math.log(curr) / Math.log(2));
    }

    static String compress(final String input) {
        final StringBuilder output = new StringBuilder();
        int sEnd = 1;
        String s = input.substring(0, sEnd);
        String c;
        while (sEnd < input.length()) {
            c = input.substring(sEnd, sEnd + 1);
            if (dict.containsKey(s + c)) {
                s = s + c;
            } else {
                output.append(dict.get(s));
                dict.put(s + c, dict.size());
                s = c;
            }
            sEnd++;
        }
        output.append(dict.get(s));
        return output.toString();
    }
}

File: src/main/java/phaserush/Mar2022/Q1.py
import math
from collections import Counter, OrderedDict
from functools import reduce


def calc_entropy(input_str):
    freq = {char: count / len(input_str) for (char, count) in Counter(input_str).items()}
    return reduce(lambda acc, curr: acc - curr * math.log2(curr), freq.values(), 0.0)


def compress(input_str, dict):
    output = ""
    sEnd = 1
    s = input_str[0: sEnd]
    while sEnd < len(input_str):
        c = input_str[sEnd: sEnd + 1]
        if s + c in dict:
            s += c
        else:
            output += str(dict[s])
            dict[s + c] = len(dict)
            s = c
        sEnd += 1
    output += str(dict[s])
    return output


if __name__ == '__main__':
    input_str = input("Enter the string to be compressed:\t")
    print(f"Entropy:\t{calc_entropy(input_str)}")
    dict = OrderedDict({"A": 0, "B": 1, "C": 2, "AB": 3, "BA": 4})
    compressed = compress(input_str, dict)
    print(f"Compressed:\t{compressed}")
    print(f"Ratio:\t{len(input_str) / len(compressed)}")
    print(f"Dictionary")
    for k, v in dict.items():
        print(k, v)
